pravega/pravega
# Issue:4818 Reader group resume from last checkpoint

Signed-off-by: Shwetha N <shwetha.n1@dell.com>

**Change log description**  
Allow passing a 'null' for a ReaderGroupConfig in resetReaderGroup() which reset the reader group to last completed checkpoint.

**Purpose of the change**  
Fixing #4818 

**What the code does**  
If ReaderGroupConfig parameter passed to resetReaderGroup()(method in ReaderGroupImpl) is null then get the last completed checkpoint and  latest reader group config from reader group state object and reset the reader group to this position. 
If there are no checkpoint triggered earlier then reset reader group to start of stream. Get the latest reader group config and stream names from reader group state object.

Below 2 scenarios are covered as part of test.
1. When application request read from A to B and there was resetReaderGroup(null) call before completion of any checkpoint then  new group should start reading from Point A not from start of stream.
2. Let us consider the case where readers in RG is reading stream s1 and created checkpoint cp1 at position p1 in the stream.

Now Stream s2 added so that readers in ReaderGroup RG starts reading stream s2 along with stream s1. This can be done by calling resetReaderGroup by passing readergroup config(config has Stream s2[pointing to head of stream] and s1 [streamcut sc1 with poistion p2 >p1, p1 is the position where checkpoint created]).

```
 newconfig = ReaderGroupConfig.builder()
                .disableAutomaticCheckpoints()
                .stream("scope/s1", startStreamCut)
                .stream("scope/s2")
                .build();
```
			
This clears all the checkpoints created in stream s1 prior to StreamCut(startStreamCut in above code).
Now Readers in RG are reading data from both streams and there is call triggered for resetReaderGroup(null) then readers in reader group reset back to newConfig ie starts reading from stream s1(streamcut sc1 with position p2) and strem s2( head of stream).


**How to verify it**  
All the testcases should pass.

CheckpointTest.java covers more scenario based testcases.
 1. testCPAndRestoreWithAddDeleteStream()  --- > Covers scenario 2 where RG reset to add new stream and resetRedaerConfig(null) triggered.
 2. testCPAndRestoreToStreamCutNoLastCheckpoint() ---- > Cover Scenario 1 where no check point created and reader group was reading from specific streamcut. Now resetRedaerConfig(null) call gets triggered.
 3. testCheckpointAndRestoreNoLastCheckpoint() ---- > Scenario where when there no check point(there was silent checkpoint created which is not considered) and reader group is reading from head of stream and resetRedaerConfig(null) call gets triggered.
 4. testCheckpointAndRestoreToLastCheckpoint() -- > Scenario where there are multiple check points and resetReaderGroup(null) starts reading from last checkpoint
